plugin/checkout_handlers.go

```diff
@@ -52,6 +53,76 @@ type CheckOutHandler interface {
 	Delete(ctx context.Context, storage logical.Storage, serviceAccountName string) error
 }
 
+// NewServiceAccountLocker is the preferable way to instantiate a ServiceAccountLocker
+// because it populates the map of locks for you.
+func NewServiceAccountLocker(child CheckOutHandler) *ServiceAccountLocker {
+	return &ServiceAccountLocker{
+		locks: &sync.Map{},
+		child: child,
+	}
+}
+
+// ServiceAccountLocker protects against races.
+type ServiceAccountLocker struct {
+	// This is, in effect, being used as a map[string]*sync.RWMutex
+	locks *sync.Map
```

Comment:
Could we reuse `locksutil` here?  It provides the same functionality from what I can see

- https://github.com/hashicorp/vault/blob/961f4468385aa08848baee1b27e64b93904c8bd7/sdk/helper/locksutil/locks.go

Example usage in the database backend: 

https://github.com/hashicorp/vault/blob/961f4468385aa08848baee1b27e64b93904c8bd7/builtin/logical/database/path_roles.go#L407-L411

Reply:
Maybe, I'm open to it. What would be the advantage to using it?

The reason I chose a *sync.Map was because I liked that it was a language-level construct, which feels very safe and optimized to me, and also because [its docs](https://golang.org/pkg/sync/#Map) describe it as being optimized for "when the entry for a given key is only ever written once but read many times", which is what is happening here.

Reply:
The advantage is simply that `locksutil` is a helper package for doing this kind of sync map work and it's use is prevalent throughout Vault's codebase. That said, this is not the Vault codebase so it would be another dependency, and it's possible that `lockutil` it only exists because it was needed in back in July 2016 when it was added to Vault, and `sync.Map` itself wasn't added to the Go language until August 2017 in go1.9.

Reply:
That said, I'll resolve this

Reply:
I was going to leave a similar comment. My thoughts on it are that `sync.Map` isn't seen all that often probably due to the comment in the docs, "Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content."

It does seem like `getOrCreateLock` is reimplementing `locksutil` to some degree. One more specific difference though is that `getOrCreateLock` isn't bounded--locks are never deleted (unless they are somewhere I've not visited yet!), and the number of locks increases with service accounts names. `locksutil` has a fixed pool so that isn't an issue.

Reply:
Hm, I like the fact that it's bounded. I also wonder about its behavior with >256 locks. [It's tests](https://github.com/hashicorp/vault/blob/master/sdk/helper/locksutil/locks_test.go) are a little light. I will test it out and see how it would work here, and also give some thought to whether it's possible that >256 service accounts might be acted upon at once.

Reply:
I actually get some unexpected behavior from that lock. This failure happens intermittently:
```
func TestManyLocks(t *testing.T) {
	r := rand.New(rand.NewSource(time.Now().UnixNano()))
	locks := CreateLocks()
	for i := 0; i < 100; i++ {
		lockName := fmt.Sprintf("%d", r.Int())
		lock := LockForKey(locks, lockName)
		lock.Lock()
		otherLock := LockForKey(locks, lockName)
		if !reflect.DeepEqual(lock, otherLock) {
			t.Fatal("they're not the same")
		}
	}
}
```
```
GOROOT=/usr/local/go #gosetup
GOPATH=/home/tbex/go #gosetup
/usr/local/go/bin/go test -c -o /tmp/___TestManyLocks_in_github_com_hashicorp_vault_sdk_helper_locksutil github.com/hashicorp/vault/sdk/helper/locksutil #gosetup
/usr/local/go/bin/go tool test2json -t /tmp/___TestManyLocks_in_github_com_hashicorp_vault_sdk_helper_locksutil -test.v -test.run ^TestManyLocks$ #gosetup
=== RUN   TestManyLocks
fatal error: all goroutines are asleep - deadlock!

goroutine 1 [chan receive]:
testing.(*T).Run(0xc0000bc100, 0x54488b, 0xd, 0x54c280, 0x469601)
	/usr/local/go/src/testing/testing.go:917 +0x381
testing.runTests.func1(0xc0000bc000)
	/usr/local/go/src/testing/testing.go:1157 +0x78
testing.tRunner(0xc0000bc000, 0xc000053e30)
	/usr/local/go/src/testing/testing.go:865 +0xc0
testing.runTests(0xc00000c080, 0x637ba0, 0x2, 0x2, 0x0)
	/usr/local/go/src/testing/testing.go:1155 +0x2a9
testing.(*M).Run(0xc0000ac000, 0x0)
	/usr/local/go/src/testing/testing.go:1072 +0x162
main.main()
	_testmain.go:44 +0x13e

goroutine 5 [semacquire]:
sync.runtime_SemacquireMutex(0xc000018fa4, 0x4f7000)
	/usr/local/go/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc000018fa0)
	/usr/local/go/src/sync/mutex.go:134 +0x109
sync.(*RWMutex).Lock(0xc000018fa0)
	/usr/local/go/src/sync/rwmutex.go:93 +0x2d
github.com/hashicorp/vault/sdk/helper/locksutil.TestManyLocks(0xc0000bc100)
	/home/tbex/go/src/github.com/hashicorp/vault/sdk/helper/locksutil/locks_test.go:24 +0x279
testing.tRunner(0xc0000bc100, 0x54c280)
	/usr/local/go/src/testing/testing.go:865 +0xc0
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:916 +0x35a

Process finished with exit code 1
```

Reply:
> it's possible that > 256 service accounts
I'm understanding the number of simultaneously held locks is just to increase performance, right? e.g. the ServiceAccountLocker _could_ just have a single lock and would behave correctly, though there would be more  lock contention for things that aren't actually at risk of conflicting (separate service accounts).

re: locks_util, yeah there could be some test additions there 🙂. The intermittent test failure above doesn't surprise me though, since you're no unlocking anything. e.g. two iterations happened to hash to the same lock ID, the second one will block since the lock is still in use.

Reply:
If the main concern is growth, another alternative might be to use `go-cache` where I can expire items after a particular amount of time, combined with a mutex if it's not thread-safe out of the box. https://github.com/patrickmn/go-cache. It has lots of test coverage and is also used in Vault. I'd be more comfortable with that.

Reply:
Hmmm, so the trade-off would be that on one hand memory is very bounded, but on the other hand there would be more resource contention. It's too bad we don't have any data on how many service accounts people tend to configure, and how quickly they will share them.

I'm thinking it would be fine to use the `locksutil` in the first iteration, especially since AD is eventually consistent anyways. Also, because a human checkout is what we expect most of the time. If we end up with problems that look like a lot of resource contention, we can always swap it out to optimize it. 

I'll go ahead and switch over.